pinot-broker/src/main/java/org/apache/pinot/broker/cursors/FsResponseStore.java

```diff
@@ -0,0 +1,280 @@
+  public static final String DATA_DIR = "data.dir";
+  public static final String FILE_NAME_EXTENSION = "extension";
+  public static final String DEFAULT_ROOT_DIR = System.getProperty("java.io.tmpdir") + File.separator + "broker"
+      + File.separator + "result_store" + File.separator;
+  public static final String DEFAULT_SCHEME = "file://";
+  public static final String DEFAULT_TEMP_DIR = DEFAULT_ROOT_DIR + "tmp";
+  public static final String DEFAULT_DATA_DIR = DEFAULT_SCHEME + DEFAULT_ROOT_DIR + "data";
+  public static final String DEFAULT_FILE_NAME_EXTENSION = "json";
+
+  private Path _localTempDir;
+  private URI _dataDir;
+  private BrokerMetrics _brokerMetrics;
+  private JsonResponseSerde _responseSerde;
+  private String _brokerHost;
+  private int _brokerPort;
+  private long _expirationIntervalInMs;
+  private String _fileExtension;
+
+  private static Path getTempPath(Path localTempDir, String... nameParts) {
+    StringBuilder filename = new StringBuilder();
+    for (String part : nameParts) {
+      filename.append(part).append("_");
+    }
+    filename.append(Thread.currentThread().getId());
+    return localTempDir.resolve(filename.toString());
+  }
+
+  private static URI combinePath(URI baseUri, String path)
+      throws URISyntaxException {
+    String newPath =
+        baseUri.getPath().endsWith(URI_SEPARATOR) ? baseUri.getPath() + path : baseUri.getPath() + URI_SEPARATOR + path;
+    return new URI(baseUri.getScheme(), baseUri.getHost(), newPath, null);
+  }
+
+  @Override
+  public String getType() {
+    return TYPE;
+  }
+
+  @Override
+  public void init(@NotNull PinotConfiguration config, @NotNull String brokerHost, int brokerPort,
+      @NotNull BrokerMetrics brokerMetrics, String expirationTime)
+      throws Exception {
+    _brokerMetrics = brokerMetrics;
+    _responseSerde = new JsonResponseSerde();
+    _brokerHost = brokerHost;
+    _brokerPort = brokerPort;
+    _expirationIntervalInMs = TimeUtils.convertPeriodToMillis(expirationTime);
+
+    _fileExtension = config.getProperty(FILE_NAME_EXTENSION, DEFAULT_FILE_NAME_EXTENSION);
+    _localTempDir = Paths.get(config.getProperty(TEMP_DIR, DEFAULT_TEMP_DIR));
+    Files.createDirectories(_localTempDir);
+
+    _dataDir = new URI(config.getProperty(DATA_DIR, DEFAULT_DATA_DIR));
+    PinotFS pinotFS = PinotFSFactory.create(_dataDir.getScheme());
+    pinotFS.mkdir(_dataDir);
+  }
+
+  @NotNull
+  @Override
+  protected BrokerMetrics getBrokerMetrics() {
+    return _brokerMetrics;
+  }
+
+  @Override
+  protected String getBrokerHost() {
+    return _brokerHost;
+  }
+
+  @Override
+  protected int getBrokerPort() {
+    return _brokerPort;
+  }
+
+  @Override
+  protected long getExpirationIntervalInMs() {
+    return _expirationIntervalInMs;
+  }
+
+  @Override
+  public boolean exists(String requestId)
+      throws Exception {
+    try (PinotFS pinotFS = PinotFSFactory.create(_dataDir.getScheme())) {
```

Comment:
This is the wrong pattern. There is a single `PinotFs` object in a hash map in `PinotFsFactory` which is re-used. `PinotFs.close` is called in `PinotFsFactory.shutdown` which is called in `**Starter.stop`. I've remove the try-catch block.

Reply:
That seems like a poor pattern since the ownership model is muddied for the closeable interface (which is why it seems like the `NoClosePinotFS` class was introduced which muddies things further IMO). However, looks like this is an established pattern of usage so changing that is outside the scope of this PR.